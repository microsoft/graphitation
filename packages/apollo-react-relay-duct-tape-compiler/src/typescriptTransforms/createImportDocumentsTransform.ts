/**
 * TODO:
 * - Don't hardcode identifiers but check if they're aliased from their imports.
 * - Support mutations and subscriptions
 * - Properly emit errors from transformer, is invariant ok?
 */

import invariant from "invariant";
import * as ts from "typescript";
import {
  FragmentDefinitionNode,
  parse as parseGraphQL,
  StringValueNode,
} from "graphql";
import * as path from "path";
import * as fs from "fs";

const PREFIX = "__graphitation_";
const QUERIES_NAMESPACE = "generatedQueries";

export function createImportDocumentsTransform(): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
    const imports: ts.Statement[] = [];

    const visitor: ts.Visitor = (node: ts.Node): ts.VisitResult<ts.Node> => {
      if (
        ts.isTaggedTemplateExpression(node) &&
        ts.isIdentifier(node.tag) &&
        node.tag.escapedText === "graphql"
      ) {
        const documentNodes = createGraphQLDocumentNodes(
          node,
          context.getCompilerOptions().module,
        );
        if (documentNodes) {
          const [modulePath, importStatement, replacementNode] = documentNodes;
          // Because we currently only emit new watch queries for fragments
          // on Node types, we cannot just assume the artefact exists for
          // every fragment definition. So check if it exists before emitting
          // the import.
          //
          // TODO: This file checking should not exist and is probably not
          //       performant.
          const artefactFile =
            path.join(
              path.dirname(node.getSourceFile().fileName),
              modulePath.text,
            ) + ".ts";
          const emitImport = fs.existsSync(artefactFile);
          if (emitImport) {
            imports.push(importStatement);
            return replacementNode;
          }
        }
      }
      return ts.visitEachChild(node, visitor, context);
    };

    return (sourceFile: ts.SourceFile) => {
      const outputSourceFile = ts.visitEachChild(sourceFile, visitor, context);
      return ts.factory.updateSourceFile(outputSourceFile, [
        ...imports,
        ...outputSourceFile.statements,
      ]);
    };
  };
}

// This is for ts-jest
export { createImportDocumentsTransform as factory };

function createGraphQLDocumentNodes(
  graphqlTagTemplateNode: ts.TaggedTemplateExpression,
  moduleKind: ts.ModuleKind = ts.ModuleKind.ES2015,
):
  | [
      modulePath: ts.StringLiteral,
      importStatement: ts.Statement,
      importReference: ts.Identifier,
    ]
  | undefined {
  const graphqlDoc = ts.isNoSubstitutionTemplateLiteral(
    graphqlTagTemplateNode.template,
  )
    ? graphqlTagTemplateNode.template.rawText
    : graphqlTagTemplateNode.template.head.rawText;
  invariant(graphqlDoc, "Expected a GraphQL document");
  const graphqlAST = parseGraphQL(graphqlDoc);
  const definitionNode = graphqlAST.definitions[0];
  if (definitionNode.kind === "OperationDefinition") {
    if (definitionNode.operation !== "query") {
      return undefined;
    }
    const operationName = definitionNode.name?.value;
    invariant(operationName, "Operations are required to have a name");
    const namespaceName = `${PREFIX}${QUERIES_NAMESPACE}_${operationName}`;
    const modulePath = createModulePathNode(operationName);
    return [
      modulePath,
      createImportStatement(moduleKind, namespaceName, modulePath),
      ts.factory.createIdentifier(namespaceName),
    ];
  } else if (definitionNode.kind === "FragmentDefinition") {
    const queryName = getQueryName(definitionNode);
    const fragmentName = definitionNode.name.value;
    const namespaceName = `${PREFIX}${QUERIES_NAMESPACE}_${fragmentName}`;
    const modulePath = createModulePathNode(queryName);
    return [
      modulePath,
      createImportStatement(moduleKind, namespaceName, modulePath),
      ts.factory.createIdentifier(namespaceName),
    ];
  }
  invariant(false, `Unhandled GraphQL definition type: ${definitionNode.kind}`);
}

function createModulePathNode(baseName: string) {
  return ts.factory.createStringLiteral(`./__generated__/${baseName}.graphql`);
}

function createImportStatement(
  moduleKind: ts.ModuleKind,
  namespaceName: string,
  modulePath: ts.StringLiteral,
): ts.Statement {
  if (moduleKind === ts.ModuleKind.CommonJS) {
    return ts.factory.createVariableStatement(
      undefined,
      ts.factory.createVariableDeclarationList([
        ts.factory.createVariableDeclaration(
          namespaceName,
          undefined,
          undefined,
          ts.factory.createPropertyAccessExpression(
            ts.factory.createCallExpression(
              ts.factory.createIdentifier("require"),
              [],
              [modulePath],
            ),
            ts.factory.createIdentifier("documents"),
          ),
        ),
      ]),
    );
  } else {
    return ts.factory.createImportDeclaration(
      undefined,
      ts.factory.createImportClause(
        false,
        undefined,
        ts.factory.createNamedImports([
          ts.factory.createImportSpecifier(
            false,
            ts.factory.createIdentifier("documents"),
            ts.factory.createIdentifier(namespaceName),
          ),
        ]),
      ),
      modulePath,
    );
  }
}

function getQueryName(definitionNode: FragmentDefinitionNode) {
  const refetchableQueryNameNode = definitionNode.directives
    ?.find((directive) => directive.name.value === "refetchable")
    ?.arguments?.find((arg) => arg.name.value === "queryName")?.value as
    | StringValueNode
    | undefined;
  if (refetchableQueryNameNode) {
    return refetchableQueryNameNode.value;
  } else {
    const fragmentName = definitionNode.name.value;
    const fragmentBaseName = fragmentName.replace(/Fragment$/, "");
    return `${fragmentBaseName}WatchNodeQuery`;
  }
}
