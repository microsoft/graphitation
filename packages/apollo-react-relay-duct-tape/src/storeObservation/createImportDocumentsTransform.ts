/**
 * TODO:
 * - Don't hardcode identifiers but check if they're aliased from their imports.
 * - Support mutations and subscriptions
 * - Properly emit errors from transformer, is invariant ok?
 */

import invariant from "invariant";
import * as ts from "typescript";
import { parse as parseGraphQL } from "graphql";
import * as path from "path";
import * as fs from "fs";

const PREFIX = "__graphitation_";
const QUERIES_NAMESPACE = "generatedQueries";

export function createImportDocumentsTransform(): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
    const imports: ts.ImportDeclaration[] = [];

    const visitor: ts.Visitor = (node: ts.Node): ts.VisitResult<ts.Node> => {
      if (
        ts.isTaggedTemplateExpression(node) &&
        ts.isIdentifier(node.tag) &&
        node.tag.escapedText === "graphql"
      ) {
        const documentNodes = createGraphQLDocumentNodes(node);
        if (documentNodes) {
          const [importDeclaration, replacementNode] = documentNodes;
          // TODO: This file checking should not exist and is probably not performant.
          const artefactFile =
            path.join(
              path.dirname(node.getSourceFile().fileName),
              (importDeclaration.moduleSpecifier as ts.StringLiteral).text
            ) + ".ts";
          if (fs.existsSync(artefactFile)) {
            imports.push(importDeclaration);
            return replacementNode;
          }
        }
      }
      return ts.visitEachChild(node, visitor, context);
    };

    return (sourceFile: ts.SourceFile) => {
      const outputSourceFile = ts.visitNode(sourceFile, visitor);
      return ts.factory.updateSourceFile(outputSourceFile, [
        ...imports,
        ...outputSourceFile.statements,
      ]);
    };
  };
}

function createGraphQLDocumentNodes(
  graphqlTagTemplateNode: ts.TaggedTemplateExpression
): [ts.ImportDeclaration, ts.Identifier] | undefined {
  const graphqlDoc = ts.isNoSubstitutionTemplateLiteral(
    graphqlTagTemplateNode.template
  )
    ? graphqlTagTemplateNode.template.rawText
    : graphqlTagTemplateNode.template.head.rawText;
  invariant(graphqlDoc, "Expected a GraphQL document");
  const graphqlAST = parseGraphQL(graphqlDoc);
  const definitionNode = graphqlAST.definitions[0];
  if (definitionNode.kind === "OperationDefinition") {
    if (definitionNode.operation !== "query") {
      return undefined;
    }
    const operationName = definitionNode.name?.value;
    invariant(operationName, "Operations are required to have a name");
    const namespaceName = `${PREFIX}${QUERIES_NAMESPACE}_${operationName}`;
    return [
      ts.factory.createImportDeclaration(
        undefined,
        undefined,
        ts.factory.createImportClause(
          false,
          undefined,
          ts.factory.createNamespaceImport(
            ts.factory.createIdentifier(namespaceName)
          )
        ),
        ts.factory.createStringLiteral(
          `./__generated__/${operationName}.graphql`
        )
      ),
      ts.factory.createIdentifier(namespaceName),
    ];
  } else if (definitionNode.kind === "FragmentDefinition") {
    const fragmentName = definitionNode.name.value;
    const fragmentBaseName = fragmentName.replace(/Fragment$/, "");
    const namespaceName = `${PREFIX}${QUERIES_NAMESPACE}_${fragmentName}`;
    return [
      ts.factory.createImportDeclaration(
        undefined,
        undefined,
        ts.factory.createImportClause(
          false,
          undefined,
          ts.factory.createNamespaceImport(
            ts.factory.createIdentifier(namespaceName)
          )
        ),
        ts.factory.createStringLiteral(
          `./__generated__/${fragmentBaseName}WatchNodeQuery.graphql`
        )
      ),
      ts.factory.createIdentifier(namespaceName),
    ];
  }
  invariant(false, `Unhandled GraphQL definition type: ${definitionNode.kind}`);
}
