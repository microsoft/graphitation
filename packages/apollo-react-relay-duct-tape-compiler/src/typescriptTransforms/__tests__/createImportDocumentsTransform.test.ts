import * as ts from "typescript";
import { createImportDocumentsTransform } from "../createImportDocumentsTransform";

import * as _fs from "fs";
const fs = _fs as jest.Mocked<typeof _fs>;

jest.mock("fs");

expect.addSnapshotSerializer({
  test: (val) => typeof val === "string",
  print: (val) => {
    return (val as string).trimRight();
  },
});

describe(createImportDocumentsTransform, () => {
  describe.each([
    [
      "es6",
      {
        target: ts.ScriptTarget.ES2015,
        module: ts.ModuleKind.ES2015,
      },
    ],
    [
      "es5",
      {
        target: ts.ScriptTarget.ES5,
        module: ts.ModuleKind.CommonJS,
      },
    ],
  ])("when emitting %s modules", (_targetName, compilerOptions) => {
    function transform(sourceText: string): string {
      return ts.transpileModule(sourceText, {
        fileName: __filename,
        transformers: { before: [createImportDocumentsTransform()] },
        compilerOptions,
      }).outputText;
    }

    it("works with documents without interpolation", () => {
      fs.existsSync.mockImplementation(() => true);
      const source = `
        const doc = graphql\`
          fragment SomeComponentFragment on Query {
            helloWorld
          }
        \`
      `;
      expect(transform(source)).toMatchSnapshot();
    });

    it("works with documents with interpolation", () => {
      fs.existsSync.mockImplementation(() => true);
      const source = `
        const doc = graphql\`
          query SomeComponentQuery($id: ID!) {
            helloWorld
            ...SomeOtherComponentFragment
          }
          \${SomeOtherComponentFragment}
        \`
      `;
      expect(transform(source)).toMatchSnapshot();
    });

    it("does not emit an import for a fragment on non-Node/Query types (for which no artefacts are emitted)", () => {
      fs.existsSync.mockImplementation(() => false);
      const source = `
      const doc = graphql\`
        fragment SomeComponentFragment on Query {
          helloWorld
        }
      \`
    `;
      expect(transform(source)).toMatchSnapshot();
    });
  });
});
