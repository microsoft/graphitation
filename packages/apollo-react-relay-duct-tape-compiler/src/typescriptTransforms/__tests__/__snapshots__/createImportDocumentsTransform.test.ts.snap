// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createImportDocumentsTransform when emitting es5 modules does not emit an import for a fragment on non-Node/Query types (for which no artefacts are emitted) 1`] = `
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var doc = graphql(__makeTemplateObject(["\\n        fragment SomeComponentFragment on Query {\\n          helloWorld\\n        }\\n      "], ["\\n        fragment SomeComponentFragment on Query {\\n          helloWorld\\n        }\\n      "]));
`;

exports[`createImportDocumentsTransform when emitting es5 modules works with documents with interpolation 1`] = `
var __graphitation_generatedQueries_SomeComponentQuery = require("./__generated__/SomeComponentQuery.graphql").documents;
var doc = __graphitation_generatedQueries_SomeComponentQuery;
`;

exports[`createImportDocumentsTransform when emitting es5 modules works with documents without interpolation 1`] = `
var __graphitation_generatedQueries_SomeComponentFragment = require("./__generated__/SomeComponentWatchNodeQuery.graphql").documents;
var doc = __graphitation_generatedQueries_SomeComponentFragment;
`;

exports[`createImportDocumentsTransform when emitting es6 modules does not emit an import for a fragment on non-Node/Query types (for which no artefacts are emitted) 1`] = `
const doc = graphql \`
        fragment SomeComponentFragment on Query {
          helloWorld
        }
      \`;
`;

exports[`createImportDocumentsTransform when emitting es6 modules works with documents with interpolation 1`] = `
import { documents as __graphitation_generatedQueries_SomeComponentQuery } from "./__generated__/SomeComponentQuery.graphql";
const doc = __graphitation_generatedQueries_SomeComponentQuery;
`;

exports[`createImportDocumentsTransform when emitting es6 modules works with documents without interpolation 1`] = `
import { documents as __graphitation_generatedQueries_SomeComponentFragment } from "./__generated__/SomeComponentWatchNodeQuery.graphql";
const doc = __graphitation_generatedQueries_SomeComponentFragment;
`;
